gcc/rust/checks/errors/rust-ast-validation.h

```diff
@@ -0,0 +1,208 @@
+  void visit (AST::MacroRulesDefinition &rules_def) override {}
+  void visit (AST::MacroInvocation &macro_invoc) override {}
+  void visit (AST::MetaItemPath &meta_item) override {}
+  void visit (AST::MetaItemSeq &meta_item) override {}
+  void visit (AST::MetaWord &meta_item) override {}
+  void visit (AST::MetaNameValueStr &meta_item) override {}
+  void visit (AST::MetaListPaths &meta_item) override {}
+  void visit (AST::MetaListNameValueStr &meta_item) override {}
+  void visit (AST::LiteralPattern &pattern) override {}
+  void visit (AST::IdentifierPattern &pattern) override {}
+  void visit (AST::WildcardPattern &pattern) override {}
+  void visit (AST::RestPattern &pattern) override {}
+  void visit (AST::RangePatternBoundLiteral &bound) override {}
+  void visit (AST::RangePatternBoundPath &bound) override {}
+  void visit (AST::RangePatternBoundQualPath &bound) override {}
+  void visit (AST::RangePattern &pattern) override {}
+  void visit (AST::ReferencePattern &pattern) override {}
+  void visit (AST::StructPatternFieldTuplePat &field) override {}
+  void visit (AST::StructPatternFieldIdentPat &field) override {}
+  void visit (AST::StructPatternFieldIdent &field) override {}
+  void visit (AST::StructPattern &pattern) override {}
+  void visit (AST::TupleStructItemsNoRange &tuple_items) override {}
+  void visit (AST::TupleStructItemsRange &tuple_items) override {}
+  void visit (AST::TupleStructPattern &pattern) override {}
+  void visit (AST::TuplePatternItemsMultiple &tuple_items) override {}
+  void visit (AST::TuplePatternItemsRanged &tuple_items) override {}
+  void visit (AST::TuplePattern &pattern) override {}
+  void visit (AST::GroupedPattern &pattern) override {}
+  void visit (AST::SlicePattern &pattern) override {}
+  void visit (AST::AltPattern &pattern) override {}
+  void visit (AST::EmptyStmt &stmt) override {}
+  void visit (AST::LetStmt &stmt) override {}
+  void visit (AST::ExprStmt &stmt) override {}
+  void visit (AST::TraitBound &bound) override {}
+  void visit (AST::ImplTraitType &type) override {}
+  void visit (AST::TraitObjectType &type) override {}
+  void visit (AST::ParenthesisedType &type) override {}
+  void visit (AST::ImplTraitTypeOneBound &type) override {}
+  void visit (AST::TraitObjectTypeOneBound &type) override {}
+  void visit (AST::TupleType &type) override {}
+  void visit (AST::NeverType &type) override {}
+  void visit (AST::RawPointerType &type) override {}
+  void visit (AST::ReferenceType &type) override {}
+  void visit (AST::ArrayType &type) override {}
+  void visit (AST::SliceType &type) override {}
+  void visit (AST::InferredType &type) override {}
+  void visit (AST::BareFunctionType &type) override {}
+
+private:
+  enum class Context
+  {
+    INHERENT_IMPL,
+    TRAIT_IMPL,
+    CRATE,
+  };
+
+  std::vector<Context> context;
+
+  void push_context (Context ctx) { context.push_back (ctx); }
+
+  void pop_context () { context.pop_back (); }
```

Comment:
I think we have a similar class in many other visitors, might be worth factoring it into a common thing. I think the macro expansion code definitely uses something like that

Reply:
I've made a whole new class for this, I'm still unsure it fits the macro visitors since the latter miss some information. In either case I'm not modifying the macro visitors in this PR, if it fits then I'll refactor it in another one.